models/llms/replit.cc

```diff
@@ -0,0 +1,726 @@
+      ggml_free(model_.ctx);
+    }
+  }
+
+  std::vector<gpt_vocab::id> Tokenize(const std::string &text) const override
+  {
+    // tokenize the prompt
+    std::vector<gpt_vocab::id> embd_inp = replit_tokenizer_tokenize(replit_tokenizer_, text);
+
+    return embd_inp;
+  }
+
+  const std::string &Detokenize(const gpt_vocab::id id) const override
+  {
+
+    const auto it = replit_tokenizer_.raw_vocab.id_to_token.find(id);
+    if (it == replit_tokenizer_.raw_vocab.id_to_token.end())
+    {
+      return kEmptyString;
+    }
+
+    current_word_ = replace_all(replit_tokenizer_.raw_vocab.id_to_token.at(id), ws_symbol, " ");
+
+    return current_word_;
+  }
+
+  bool IsEosToken(const gpt_vocab::id token) const override
+  {
+    if (token == EosToken())
+    {
+      return true;
+    }
+    // Handle special tokens in StarChat and Dolly V2.
+    if (!replit_tokenizer_.raw_vocab.special_tokens.empty())
+    {
+      const std::string &text = Detokenize(token);
+      return text == "<|end|>" || text == "### End";
+    }
+    return false;
+  }
+
+  gpt_vocab::id EosToken() const override
+  {
+    const auto it = replit_tokenizer_.raw_vocab.token_to_id.find("<|endoftext|>");
+    if (it != replit_tokenizer_.raw_vocab.token_to_id.end())
+    {
+      return it->second;
+    }
+    return 0;
+  }
+
+  int VocabSize() const override { return replit_tokenizer_.raw_vocab.id_to_token.size(); }
+
+  gpt_vocab::id Sample(const int top_k, const float top_p,
+                       const float temperature,
+                       const float repetition_penalty,
+                       int last_n_tokens, int seed) const override
+  {
+    if (logits_.empty())
+    {
+      return EosToken();
+    }
+    if (last_n_tokens < 0)
+    {
+      last_n_tokens = ContextLength();
+    }
+    if (seed < 0)
+    {
+      seed = time(nullptr);
+    }
+    std::mt19937 rng(seed);
+
+    std::unordered_set<gpt_vocab::id> recent_tokens;
+    if (repetition_penalty != 1.0f)
+    {
+      recent_tokens = previous_tokens_.GetRecent(last_n_tokens);
+    }
+
+    return gpt_sample_top_k_top_p(
+        replit_tokenizer_.raw_vocab, logits_.data() + (logits_.size() - VocabSize()), top_k, top_p,
+        temperature, repetition_penalty, recent_tokens, rng);
+  }
+
+protected:
+  replit_tokenizer replit_tokenizer_;
+  bool Load(const std::string &filename, const int context_length,
+            const int gpu_layers) override
+  {
+    if (context_length > 0)
+    {
+      model_.hparams.n_ctx = context_length;
+    }
+    if (!replit_model_load(filename, model_, replit_tokenizer_))
+    {
+      return false;
+    }
+    n_ctx_ = model_.hparams.n_ctx;
+    return true;
+  }
+
+  bool Eval(const std::vector<gpt_vocab::id> &tokens, const int threads,
+            const int n_past) override
+  {
+    return replit_eval(model_, threads, n_past, tokens, logits_,
+                       mem_per_token_);
+  }
+
+private:
+  replit_model model_;
+  mutable std::string current_word_;
```

Comment:
Nit: You can call it something like `detokenized_text_` to make it easier to understand that it is used by detokenize method.